README.md
# 🏃‍♀️ `ptr` - Python Test Runner 🏃‍♂️

[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/ambv/black)
[![Actions Status](https://github.com/facebookincubator/ptr/workflows/ptr_ci/badge.svg)](https://github.com/facebookincubator/ptr/actions)
[![PyPI](https://img.shields.io/pypi/v/ptr)](https://pypi.org/project/ptr/)
[![Downloads](https://pepy.tech/badge/ptr/week)](https://pepy.tech/project/ptr/week)

Python Test Runner (ptr) was born to run tests in an opinionated way, within arbitrary code repositories.
`ptr` supports many Python projects with unit tests defined in their `setup.(cfg|py)` files per repository.
`ptr` allows developers to test multiple projects/modules in one Python environment through the use of a single test virtual environment.

- `ptr` requires `>=` **python 3.7**
- `ptr` itself uses `ptr` to run its tests 👌🏼
- `ptr` is supported and tested on *Linux*, *MacOS* + *Windows* Operating Systems

By adding `ptr` configuration to your either of your `pyproject.toml`, `setup.cfg` or `setup.py` you can have `ptr` perform the following,
per test suite, in parallel:

- run your test suite
- check and enforce coverage requirements (via [coverage](https://pypi.org/project/coverage/)),
- format code (via [black](https://pypi.org/project/black/))
- perform static type analysis (via [mypy](http://mypy-lang.org/))


## Quickstart
- Install `ptr` into you virtualenv
  - `pip install ptr`
- Ensure your tests have a base file that can be executed directly
  - i.e. `python3 test.py` (possibly using `unittest.main()`)
- After adding `ptr_params` to setup.py (see example below), run:
```
cd repo
ptr
```

## How does `ptr` perform this magic? 🎩

I'm glad you ask. Under the covers `ptr` performs:
- Recursively searches for `setup.(cfg|py)` files from `BASE_DIR` (defaults to your "current working directory" (CWD))
   - [AST](https://docs.python.org/3/library/ast.html) parses out the config for each `setup.py` test requirements
   - If a `pyproject.toml` or `setup.cfg` exists, load via configparser/tomli and prefer if a `[ptr]` section exists
- Creates a [Python Virtual Environment](https://docs.python.org/3/tutorial/venv.html) (*OPTIONALLY* pointed at an internal PyPI mirror)
- Runs `ATONCE` tests suites in parallel (i.e. per setup.(cfg|ptr))
- All steps will be run for each suite and ONLY *FAILED* runs will have output written to stdout

## Usage 🤓

To use `ptr` all you need to do is cd to your project or set the base dir via `-b` and execute:

    $ ptr [-dk] [-b some/path] [--venv /tmp/existing_venv]

For **faster runs** when testing, it is recommended to reuse a Virtual Environment:
- `-k` - To keep the virtualenv created by `ptr`.
- Use `--venv VENV_PATH` to reuse to an existing virtualenv created by the user.

### Help Output 🙋‍♀️ 🙋‍♂️

```shell
usage: ptr.py [-h] [-a ATONCE] [-b BASE_DIR] [-d] [-e] [-k] [-m MIRROR]
              [--print-cov] [--print-non-configured]
              [--progress-interval PROGRESS_INTERVAL] [--run-disabled]
              [--stats-file STATS_FILE] [--system-site-packages] [--venv VENV]
              [--venv-timeout VENV_TIMEOUT]

optional arguments:
  -h, --help            show this help message and exit
  -a ATONCE, --atonce ATONCE
                        How many tests to run at once [Default: 6]
  -b BASE_DIR, --base-dir BASE_DIR
                        Path to recursively look for setup.py files [Default:
                        /Users/cooper/repos/ptr]
  -d, --debug           Verbose debug output
  -e, --error-on-warnings
                        Have Python warnings raise DeprecationWarning on tests
                        run
  -k, --keep-venv       Do not remove created venv
  -m MIRROR, --mirror MIRROR
                        URL for pip to use for Simple API [Default:
                        https://pypi.org/simple/]
  --print-cov           Print modules coverage report
  --print-non-configured
                        Print modules not configured to run ptr
  --progress-interval PROGRESS_INTERVAL
                        Seconds between status update on test running
                        [Default: Disabled]
  --run-disabled        Force any disabled tests suites to run
  --stats-file STATS_FILE
                        JSON statistics file [Default: /var/folders/tc/hbwxh76
                        j1hn6gqjd2n2sjn4j9k1glp/T/ptr_stats_12510]
  --system-site-packages
                        Give the virtual environment access to the system
                        site-packages dir
  --venv VENV           Path to venv to reuse
  --venv-timeout VENV_TIMEOUT
                        Timeout in seconds for venv creation + deps install
                        [Default: 120]
```

## Configuration 🧰

`ptr` is configured by placing directives in one or more of the following files.  `.ptrconfig` provides
base configuration and default values for all projects in the repository, while each `setup.(cfg|py)`
overrides the base configuration for the respective packages they define.

### `.ptrconfig`

`ptr` supports a general config in `ini` ([ConfigParser](https://docs.python.org/3/library/configparser.html)) format.
A `.ptrconfig` file can be placed at the root of any repository or in any directory within your repository.
The first `.ptrconfig` file found via a recursive walk to the root ("/" in POSIX systems) will be used.

Please refer to [`ptrconfig.sample`](http://github.com/facebookincubator/ptr/blob/master/ptrconfig.sample) for the options available.

### `setup.py`

This is per project in your repository. A simple example, based on `ptr` itself:

```python
# Specific Python Test Runner (ptr) params for Unit Testing Enforcement
ptr_params = {
    # Where mypy will run to type check your program
    "entry_point_module": "ptr",
    # Base Unittest file
    "test_suite": "ptr_tests",
    "test_suite_timeout": 300,
    # Relative path from setup.py to module (e.g. ptr == ptr.py)
    "required_coverage": {"ptr.py": 99, "TOTAL": 99},
    # Run `black --check` or not
    "run_black": False,
    # Run mypy or not
    "run_mypy": True,
}
```

### `pyproject.toml`

This is per project in your repository and if exists is preferred over `setup.py` and `setup.cfg`.

Please refer to [`pyproject.toml`](http://github.com/facebookincubator/ptr/blob/master/pyproject.toml)
for the options available + format.

### `setup.cfg`

This is per project in your repository and if exists is preferred over `setup.py`.

Please refer to [`setup.cfg.sample`](http://github.com/facebookincubator/ptr/blob/master/setup.cfg.sample)
for the options available + format.

### mypy Specifics

When enabled, (in `setup.(cfg|py)`) **mypy** can support using a custom `mypy.ini` for each setup.py (module) defined.

To have `ptr` run mypy using you config:
- create a `mypy.ini` in the same directory as your `setup.py`
- OR add **[mypy]** section to your `setup.cfg`

`mypy` Configuration Documentation can be found [here](https://mypy.readthedocs.io/en/stable/config_file.html)
- An example `setup.cfg` can be seen [here](http://github.com/facebookincubator/ptr/blob/master/mypy.ini).

# Example Output 📝

Here are some example runs.

## Successful `ptr` Run:

Here is what you want to see in your CI logs!
```
[2019-02-06 21:51:45,442] INFO: Starting ptr.py (ptr.py:782)
[2019-02-06 21:51:59,471] INFO: Successfully created venv @ /var/folders/tc/hbwxh76j1hn6gqjd2n2sjn4j9k1glp/T/ptr_venv_24397 to run tests (14s) (ptr.py:547)
[2019-02-06 21:51:59,472] INFO: Installing /Users/cooper/repos/ptr/setup.py + deps (ptr.py:417)
[2019-02-06 21:52:00,726] INFO: Running /Users/cooper/repos/ptr/ptr_tests.py tests via coverage (ptr.py:417)
[2019-02-06 21:52:04,153] INFO: Analyzing coverage report for /Users/cooper/repos/ptr/setup.py (ptr.py:417)
[2019-02-06 21:52:04,368] INFO: Running mypy for /Users/cooper/repos/ptr/setup.py (ptr.py:417)
[2019-05-03 14:54:09,915] INFO: Running flake8 for /Users/cooper/repos/ptr/setup.py (ptr.py:417)
[2019-05-03 14:54:10,422] INFO: Running pylint for /Users/cooper/repos/ptr/setup.py (ptr.py:417)
[2019-05-03 14:54:14,020] INFO: Running pyre for /Users/cooper/repos/ptr/setup.py (ptr.py:417)
[2019-02-06 21:52:07,733] INFO: /Users/cooper/repos/ptr/setup.py has passed all configured tests (ptr.py:509)
-- Summary (total time 22s):

✅ PASS: 1
❌ FAIL: 0
⌛️ TIMEOUT: 0
💩 TOTAL: 1

-- 1 / 1 (100%) `setup.py`'s have `ptr` tests running
```
## Unsuccessful `ptr` Run Examples:

Here are some examples of runs failing. Any "step" can fail. All output is predominately the underlying tool.

### Unit Test Failure

```
[2019-02-06 21:53:58,121] INFO: Starting ptr.py (ptr.py:782)
[2019-02-06 21:53:58,143] INFO: Installing /Users/cooper/repos/ptr/setup.py + deps (ptr.py:417)
[2019-02-06 21:53:59,698] INFO: Running /Users/cooper/repos/ptr/ptr_tests.py tests via coverage (ptr.py:417)
-- Summary (total time 5s):

✅ PASS: 0
❌ FAIL: 1
⌛️ TIMEOUT: 0
💩 TOTAL: 1

-- 1 / 1 (100%) `setup.py`'s have `ptr` tests running

-- Failure Output --

/Users/cooper/repos/ptr/setup.py (failed 'tests_run' step):
...F....................
======================================================================
FAIL: test_config (__main__.TestPtr)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/cooper/repos/ptr/ptr_tests.py", line 125, in test_config
    self.assertEqual(len(sc["ptr"]["venv_pkgs"].split()), 4)
AssertionError: 5 != 4

----------------------------------------------------------------------
Ran 24 tests in 3.221s

FAILED (failures=1)
```

### coverage

```
[2019-02-06 21:55:42,947] INFO: Starting ptr.py (ptr.py:782)
[2019-02-06 21:55:42,969] INFO: Installing /Users/cooper/repos/ptr/setup.py + deps (ptr.py:417)
[2019-02-06 21:55:44,920] INFO: Running /Users/cooper/repos/ptr/ptr_tests.py tests via coverage (ptr.py:417)
[2019-02-06 21:55:49,628] INFO: Analyzing coverage report for /Users/cooper/repos/ptr/setup.py (ptr.py:417)
-- Summary (total time 7s):

✅ PASS: 0
❌ FAIL: 1
⌛️ TIMEOUT: 0
💩 TOTAL: 1

-- 1 / 1 (100%) `setup.py`'s have `ptr` tests running

-- Failure Output --

/Users/cooper/repos/ptr/setup.py (failed 'analyze_coverage' step):
The following files did not meet coverage requirements:
  ptr.py: 84 < 99 - Missing: 146-147, 175, 209, 245, 269, 288-291, 334-336, 414-415, 425-446, 466, 497, 506, 541-543, 562, 611-614, 639-688
```

### black

```
[2019-02-06 22:34:20,029] INFO: Starting ptr.py (ptr.py:804)
[2019-02-06 22:34:20,060] INFO: Installing /Users/cooper/repos/ptr/setup.py + deps (ptr.py:430)
[2019-02-06 22:34:21,614] INFO: Running /Users/cooper/repos/ptr/ptr_tests.py tests via coverage (ptr.py:430)
[2019-02-06 22:34:25,208] INFO: Analyzing coverage report for /Users/cooper/repos/ptr/setup.py (ptr.py:430)
[2019-02-06 22:34:25,450] INFO: Running mypy for /Users/cooper/repos/ptr/setup.py (ptr.py:430)
[2019-02-06 22:34:26,422] INFO: Running black for /Users/cooper/repos/ptr/setup.py (ptr.py:430)
-- Summary (total time 7s):

✅ PASS: 0
❌ FAIL: 1
⌛️ TIMEOUT: 0
💩 TOTAL: 1

-- 1 / 1 (100%) `setup.py`'s have `ptr` tests running

-- Failure Output --

/Users/cooper/repos/ptr/setup.py (failed 'black_run' step):
would reformat /Users/cooper/repos/ptr/ptr.py
All done! 💥 💔 💥
1 file would be reformatted, 4 files would be left unchanged.
```

### mypy

```
[2019-02-06 22:35:39,480] INFO: Starting ptr.py (ptr.py:802)
[2019-02-06 22:35:39,531] INFO: Installing /Users/cooper/repos/ptr/setup.py + deps (ptr.py:428)

CHANGES.md
## `ptr` Change History

Each release to PyPI I'm going to give a codename as to where I am or was in the world 🌏.

### 2022.7.12

Codename: **Seattle 🦑**

*Release named in honoUr of my visit to Seattle last week ...*

- 3.11 Supported - added metadata + made 3.11 CI pass - PR #121
- Add `pyproject.toml` user config support - PR #120
- Add basic Hypothesis fuzz testing - PR #118
  - Learnt during US PyCon 2022
- Add `pyproject.toml` basic PEP517 building support - PR #117
- This *(again)* will be the last 3.7 supported version
  - I forgot to remove + CI still passes so will leave for now :)

### 2022.2.2

Codename: **Richie 2 for 22 🏏**

*Release named in honoUr of the Late Great [Richie Benaud](https://en.wikipedia.org/wiki/Richie_Benaud)*

- This will be the last 3.7 supported version
  - Why? Want a less buggy `asyncio.run` for starting asyncio
- Reactivate `pylint` in CI 
  - f-string all ptr strings
- Drop python 3.6 CI + support
- We're Meta now - Update copyrights due to comapny change name

3.11 CI is failing due to depedencies. Will watch and fix asap.

### 2021.11.23

Codename: **A 🅰️**

*Release to support 'Three dot Ten' and A == 10 in hexidecimal*

- Add Official Support and CI for Python 3.10
  - This will be the last release *(if no bugs present themselves)* for 3.6
- Also start ci_latest running 3.11

### 2021.5.28

Codename: **Memorial Day 🇺**

*Release on Memorial Day long weekend*

- Add usort to the family of CI options
  - https://pypi.org/project/usort/

### 2021.3.16

Codename: **Wildwood 🪵**

*First release from my South Lake Tahoe residence*

- Support floats for coverage comparisions in `required_coverage` - PR #108
- Change `--print-cov` to work when `required_coverage` does not exist in ptr config - PR #109
- Move tests out of install but include in sdist - PR #99 - Thanks @jayvdb
- Test with 3.10 alphas via deadsnakes - PR #110
- Misc f-string + lint/typing fixes

### 2020.2.26

Codename: **Month after Straya Day 🇦🇺**🇺

*1 month since @cooperlees was in AU for Australia day!*

- Enable pyre for Python 3.8 now it's fully supported

### 2019.12.25

Codename: **Vernon Hills, IL 🎅**

*@cooperlees visiting girlfriend's family in IL*

- Fix documentation for some new arguments - Thanks @omikader
- Windows now defaults to updating pip + setuptools
- Rely on setting current working directory with subprocess and remove coverage file environment setting
- f-string everything via [flynt](https://pypi.org/project/flynt/) now we're >= 3.6

### 2019.12.13

Codename: **College Park, MD 🐢**

*@omikader studied at the University of Maryland, College Park*

- Add ability to create partial `.ptrconfig` files - PR: #83
- Use native recursive file search for black and flake8 - PR: #81

### 2019.12.12

Codename: **Sapporo, Japan 🇯🇵**

*@omikader is going skiing there in February*

- Add support for passing `--system-site-packages` during venv creation - PR: #80

### 2019.11.21

Codename: **Santa Clara, CA**

*@thatch lives there!**

- Add license file for @thatch - PR: #77 - *Thanks @thatch*

### 2019.11.15

Codename: **Russian River, CA 🇺🇸**

*@cooperlees is going to the Russian River tomorrow*

- Run ptr even when test_suite is not provided - PR: #73 - *Thanks @spurav*
- Remove use of deprecated setuptools `test` option
- Move to GitHub Actions for CI + Releases

### 2019.11.2

Codename: **Mumbai, India 🇮🇳**

*@spurav is from Mumbai, India*

- Fix math error dividing 1 / 2 for os.cpu_count - PR: #69 - *Thanks @spurav*
- Get GitHub Actions CI Running - PR: #68 + #70 - *Thanks @adhaamehab*

### 2019.10.22

Codename: **Cairo, Egypt 🇪🇬**

*@adhaamehab who first time contributed is from there! Thanks!*

- Print more helpful error when file does not exist for coverage check - PR: #51
- Officially Support Python 3.8 and enforce tests passing
- Disable pyre-check for >=3.8 due to no support
  - Issue to Track: https://github.com/facebook/pyre-check/issues/213

### 2019.10.6

Codename: **Stanford, CA**

*@cooperlees tailgated there yesterday 🚐🏈*

- Handle OSX /private in coverage report - PR: #60

### 2019.9.14

Codename: **Clowntown 🤡**

- Fix bad bool passing for `--print-cov` + `--error-on-warnings`

### 2019.9.11

Codename: **Rathbone Square**

*@cooperlees is releasing this release from FB London Office*

- Add CLI option to enable erroring on all [warnings](https://docs.python.org/3/library/warnings.html) - Issue: #57

### 2019.8.7

Codename: **Ellicott City**

*@omikader who reported and tested the fix for AST parsing is from there.*
*Ellicott City is also home to one of the oldest surviving train stations in the US!*

- Fix AST parsing to ignore AST targets that do not have an `id` attribute - Issue: #54
- Renable `black` by default on 3.7 now that it runs on > 3.7.2 - Issue: #41
- Fix --venv-timeout to store ints
- Refactor _get_test_modules function to remove lint error - PR: #53 - *Thanks @TomasFeeney*

### 2019.7.16

Codename: **California City**

*An over developed desert with infrastructure and little [people](https://en.wikipedia.org/wiki/California_City,_California)*

- Add `--print-non-configured` to find non `ptr` modules in repos - Issue: #50

### 2019.6.15

Codename: **Concord**

*🛫 @cooperlees is playing Aussie Football @ Concord, CA today 🏈*

- Add ability to `disable` test suites and add a `--run-disabled` option to force the run - Issue: #46
- Azure CI now runs on Mac and Windows again

### 2019.5.3

Codename: **Cleveland**

*🇺🇸 @cooperlees + @jreese are at PyCon US in Cleveland, OH 🦅*

- Print step name in failure output - Issue: #31
- Add in `pyre` Type Checking step support - Issue: #38 + #40

**Known Bug:** `black.exe` does not run in Windows 3.7 - **disabled** by *default* on Python 3.7 on Windows


ptr_tests_fixtures.py
#!/usr/bin/env python3
# Copyright © Meta Platforms, Inc. and affiliates

# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
# coding=utf8
# flake8: noqa

from os import environ
from os.path import sep
from pathlib import Path
from sys import version_info
from tempfile import gettempdir
from typing import Dict

from ptr import test_result


class FakeEventLoop:
    def close(self, *args, **kwargs) -> None:
        pass

    def run_until_complete(self, *args, **kwargs) -> int:
        return 0


# Disabled is set as we --run-disabled the run in CI
EXPECTED_TEST_PARAMS = {
    "disabled": True,
    "entry_point_module": "ptr",
    "test_suite": "ptr_tests",
    "test_suite_timeout": 120,
    "required_coverage": {"ptr.py": 84, "TOTAL": 88},
    "run_usort": True,
    "run_black": True,
    "run_mypy": True,
    "run_flake8": True,
    "run_pylint": False,
    "run_pyre": True,
}

EXPECTED_COVERAGE_FAIL_RESULT = test_result(
    setup_py_path=Path("unittest/setup.py"),
    returncode=3,
    output=(
        "The following files did not meet coverage requirements:\n"
        + f"  unittest{sep}ptr.py: 69.0 < 99.0 - Missing: 70-72, 76-94, 98\n"
    ),
    runtime=0,
    timeout=False,
)
EXPECTED_PTR_COVERAGE_FAIL_RESULT = test_result(
    setup_py_path=Path("unittest/setup.py"),
    returncode=3,
    output=(
        f"The following files did not meet coverage requirements:\n  tg{sep}tg.py: "
        + "22.0 < 99 - Missing: 39-59, 62-73, 121, 145-149, 153-225, 231-234, 238\n  "
        + "TOTAL: 40.0 < 99 - Missing: \n"
    ),
    runtime=0,
    timeout=False,
)
EXPECTED_PTR_COVERAGE_MISSING_FILE_RESULT = test_result(
    setup_py_path=Path("unittest/setup.py"),
    returncode=3,
    output=(
        "fake_file.py has not reported any coverage. Does the file exist? "
        + "Does it get ran during tests? Remove from setup config."
    ),
    runtime=0,
    timeout=False,
)

EXPECTED_COVERAGE_RESULTS = [
    test_result(
        setup_py_path=Path("project69/setup.py"),
        returncode=0,
        output="Killed it",
        runtime=4,
        timeout=False,
    ),
    test_result(
        setup_py_path=Path("project1/setup.py"),
        returncode=2,
        output="Timeout during Running project1/tests.py tests via coverage",
        runtime=1,
        timeout=True,
    ),
    test_result(
        setup_py_path=Path("project2/setup.py"),
        returncode=1,
        output="..F..\nShit Failed yo!",
        runtime=2,
        timeout=False,
    ),
]

FAKE_REQ_COVERAGE = {str(Path("unittest/ptr.py")): 99.0, "TOTAL": 99}
FAKE_TG_REQ_COVERAGE: Dict[str, float] = {str(Path("tg/tg.py")): 99, "TOTAL": 99}


SAMPLE_REPORT_OUTPUT = """\
Name                                Stmts   Miss  Cover   Missing
------------------------------------------------------------------
unittest{sep}ptr.py                     59     14     69%     70-72, 76-94, 98
unittest{sep}ptr_tests.py               24      0     100%
unittest{sep}ptr_venv_fixtures.py        1      0     100%
------------------------------------------------------------------
TOTAL                                84     14    99%
""".format(
    sep=sep
)

SAMPLE_FLOAT_REPORT_OUTPUT = """\
Name                                Stmts   Miss  Cover   Missing
------------------------------------------------------------------
unittest{sep}ptr.py                     59     14     69.00%     70-72, 76-94, 98
unittest{sep}ptr_tests.py               24      0     100.00%
unittest{sep}ptr_venv_fixtures.py        1      0     100.00%
------------------------------------------------------------------
TOTAL                                84     14    99.00%
""".format(
    sep=sep
)

HARD_SET_VENV = Path(f"{gettempdir()}/ptr_venv_2580217")
BASE_VENV_PATH = (
    Path(environ["VIRTUAL_ENV"]) if "VIRTUAL_ENV" in environ else HARD_SET_VENV
)
SAMPLE_NIX_TG_REPORT_OUTPUT = """\
Name                                                                         Stmts   Miss  Cover   Missing
----------------------------------------------------------------------------------------------------------
{base_venv_path}/lib/python{major}.{minor}/site-packages/click/__init__.py             13      0   100%
{base_venv_path}/lib/python{major}.{minor}/site-packages/click/_compat.py             433    320    26%   22, 27-31, 41-42, 47-50, 57-59, 65-76, 79-82, 86, 100-102, 105, 108-116, 119-128, 131-143, 146-153, 157-260, 275-278, 289-292, 299-307, 314-322, 333-345, 349-374, 380, 387-404, 408-412, 415-419, 422-426, 429-432, 438, 443-446, 450-455, 459-468, 475-516, 524-525, 531-534, 538, 541-550, 553, 556, 559, 562, 575-579, 587-645, 648, 659-662, 671-672, 674, 679-680
{base_venv_path}/lib/python{major}.{minor}/site-packages/click/_textwrap.py            30     16    47%   8-17, 32-38
{base_venv_path}/lib/python{major}.{minor}/site-packages/click/_unicodefun.py          71     54    24%   16-29, 34, 37-41, 53, 57-58, 62-120
{base_venv_path}/lib/python{major}.{minor}/site-packages/click/core.py                877    389    56%   35-36, 43-45, 56-58, 62-70, 81, 90-95, 107-111, 113-115, 237, 246, 262, 267, 302, 310, 328-331, 335, 351, 418, 434-435, 439-441, 453, 458-461, 465-469, 475-478, 484-488, 496, 500, 510, 534-555, 562-573, 616, 619, 637-638, 642, 649, 656, 698, 701, 706, 717-727, 729-730, 732-735, 737-742, 755-760, 764, 809, 819-821, 856, 884, 908, 910-913, 930-932, 942-948, 954-956, 995, 1005-1007, 1049-1053, 1062-1069, 1073-1082, 1086-1087, 1090-1096, 1099-1164, 1167-1190, 1196, 1202, 1221-1225, 1233-1237, 1245-1249, 1252, 1266-1268, 1272, 1275-1280, 1283-1286, 1346, 1366, 1370, 1375, 1381-1385, 1388, 1393, 1395, 1404-1411, 1416, 1431-1435, 1441, 1444, 1449-1457, 1460-1463, 1470-1473, 1482-1484, 1487, 1496-1497, 1550, 1554, 1563, 1585, 1597, 1599, 1601, 1604, 1608, 1611, 1622-1624, 1644, 1647-1649, 1652, 1666, 1689, 1695, 1707-1713, 1718-1726, 1729, 1731, 1740-1745, 1754-1761, 1767-1773, 1776-1784, 1789, 1803-1810, 1815-1817, 1820-1829, 1832-1842, 1845, 1848, 1851
{base_venv_path}/lib/python{major}.{minor}/site-packages/click/decorators.py          147     89    39%   17, 27, 53-66, 71, 77-78, 83-85, 132, 149-153, 195-205, 220-225, 240-283, 296-307
{base_venv_path}/lib/python{major}.{minor}/site-packages/click/exceptions.py          115     81    30%   6-8, 18-23, 26, 29, 32-35, 38-40, 54-56, 59-70, 93-95, 98-106, 123-124, 127-148, 165-169, 172-179, 193-194, 206, 213-218, 221
{base_venv_path}/lib/python{major}.{minor}/site-packages/click/formatting.py          134     18    87%   59, 61, 68-69, 82, 108, 143-146, 185, 192-193, 197-198, 205-208, 250
{base_venv_path}/lib/python{major}.{minor}/site-packages/click/globals.py              18      4    78%   24-26, 45
{base_venv_path}/lib/python{major}.{minor}/site-packages/click/parser.py              236    122    48%   41-47, 50-64, 69-71, 77-79, 85, 94-95, 100-113, 126, 136, 150, 153-160, 167-169, 172-180, 239, 255-257, 270-272, 280, 292, 295-299, 323-325, 333-343, 346, 354-398, 406, 416-427
{base_venv_path}/lib/python{major}.{minor}/site-packages/click/termui.py              195    165    15%   42-43, 47-52, 92-138, 160-182, 190-229, 244-258, 344-346, 361-369, 425-449, 461, 476-478, 510-515, 540-541, 569-572, 576-577, 595-606
{base_venv_path}/lib/python{major}.{minor}/site-packages/click/testing.py             207     86    58%   19, 28-29, 32, 35-36, 39, 42, 45, 48, 51, 57-62, 66-67, 69, 94, 99, 105-107, 112, 159, 191-197, 201, 206-207, 216-220, 223-225, 228-232, 239, 253-260, 264-270, 318, 331, 334, 337-346, 364-374
{base_venv_path}/lib/python{major}.{minor}/site-packages/click/types.py               318    203    36%   55, 65, 69, 77, 83-84, 87-94, 101, 104, 111-123, 126, 145-146, 149, 152, 156-179, 183, 209, 216, 219-222, 226-231, 236, 243-246, 249, 268-285, 288, 295-298, 302, 316-318, 321-338, 341, 350-355, 358, 365-371, 374, 408-412, 415-421, 424-452, 493-510, 513-518, 521-560, 578, 582, 586, 589-592, 602, 608, 610, 620, 622, 628-634
{base_venv_path}/lib/python{major}.{minor}/site-packages/click/utils.py               179    124    31%   13-14, 22, 27-32, 37-42, 47-67, 79-95, 98, 101-103, 110-121, 125-126, 132-133, 136, 139, 142-143, 149, 152, 155, 158, 161, 164, 216, 222, 229, 237-242, 253-257, 274-277, 291-294, 321-327, 346-348, 363-365, 403-414, 429, 432-437, 440
{base_venv_path}/lib/python{major}.{minor}/site-packages/tabulate.py                  545    455    17%   14-25, 41, 113-121, 127-128, 132-141, 145-148, 153, 157-170, 173-181, 184-186, 195-199, 202-216, 412, 419-423, 439-444, 454, 469, 491-510, 526-537, 547-548, 558-559, 569-570, 574, 579-582, 593-600, 604-607, 612, 617-627, 631-654, 660-689, 693-696, 721-722, 737-756, 761-775, 780-787, 792-795, 829-955, 1244-1309, 1320-1326, 1330-1335, 1340-1341, 1346-1351, 1355-1356, 1360-1369, 1374-1381, 1385-1386, 1391-1432, 1458-1503, 1508-1510, 1514
{base_venv_path}/lib/python{major}.{minor}/site-packages/tg/__init__.py                 0      0   100%
{base_venv_path}/lib/python{major}.{minor}/site-packages/tg/commands/__init__.py        0      0   100%
{base_venv_path}/lib/python{major}.{minor}/site-packages/tg/commands/base.py          228    173    24%   33, 39, 43, 61-64, 69-72, 77-81, 87-117, 121-124, 129-132, 135-148, 155-188, 193-219, 222-237, 244-279, 282-293, 296-305, 308-317, 320-331, 334-343, 347-360, 363-388, 391-415, 418-431, 434-436, 439-441
{base_venv_path}/lib/python{major}.{minor}/site-packages/tg/commands/consts.py         20      0   100%
{base_venv_path}/lib/python{major}.{minor}/site-packages/tg/commands/zeroize.py        45      1    98%   24
{base_venv_path}/lib/python{major}.{minor}/site-packages/tg/tests/base.py              26      0   100%
{base_venv_path}/lib/python{major}.{minor}/site-packages/tg/tests/test_zeroize.py      29      1    97%   44
{base_venv_path}/lib/python{major}.{minor}/site-packages/tg/tg.py                     116     90    22%   39-59, 62-73, 121, 145-149, 153-225, 231-234, 238
----------------------------------------------------------------------------------------------------------
TOTAL                                                                         3982   2391    40%
""".format(
    base_venv_path=str(BASE_VENV_PATH),
    major=version_info.major,
    minor=version_info.minor,
)

SAMPLE_WIN_TG_REPORT_OUTPUT = """\
Name                                                                         Stmts   Miss  Cover   Missing
----------------------------------------------------------------------------------------------------------
C:\\temp\\tp\\Lib\\site-packages\\click\\testing.py              207     86    58%   19, 28-29, 32, 35-36, 39, 42, 45, 48, 51, 57-62, 66-67, 69, 94, 99, 105-107, 112, 159, 191-197, 201, 206-207, 216-220, 223-225, 228-232, 239, 253-260, 264-270, 318, 331, 334, 337-346, 364-374
C:\\temp\\tp\\Lib\\site-packages\\click\\types.py                207     86    58%   19, 28-29, 32, 35-36, 39, 42, 45, 48, 51, 57-62, 66-67
{base_venv_path}\\Lib\\site-packages\\tg\\__init__.py            0      0   100%
{base_venv_path}\\Lib\\site-packages\\tg\\commands\\__init__.py  0      0   100%
{base_venv_path}\\Lib\\site-packages\\tg\\commands\\base.py      228    173    24%   33, 39, 43, 61-64, 69-72, 77-81, 87-117, 121-124, 129-132, 135-148, 155-188, 193-219, 222-237, 244-279, 282-293, 296-305, 308-317, 320-331, 334-343, 347-360, 363-388, 391-415, 418-431, 434-436, 439-441
{base_venv_path}\\Lib\\site-packages\\tg\\commands\\consts.py    20      0   100%
{base_venv_path}\\Lib\\site-packages\\tg\\commands\\zeroize.py   45      1    98%   24
{base_venv_path}\\Lib\\site-packages\\tg\\tests\\base.py         26      0   100%
{base_venv_path}\\Lib\\site-packages\\tg\\tests\\test_zeroize.py 29      1    97%   44
{base_venv_path}\\Lib\\site-packages\\tg\\tg.py                  116     90    22%   39-59, 62-73, 121, 145-149, 153-225, 231-234, 238
----------------------------------------------------------------------------------------------------------
TOTAL                                                    3982   2391    40%
""".format(
    base_venv_path=str(BASE_VENV_PATH)
)

SAMPLE_SETUP_PY = """\
#!/usr/bin/env python3
# Copyright © Meta Platforms, Inc

from setuptools import setup


# Specific Python Test Runner (ptr) params for Unit Testing Enforcement
ptr_params = {
    "test_suite": "coop.tests.base",
    "test_suite_timeout": 60,
    "required_coverage": {
        "coop/coop.py": 99,
        "coop/commands/pwn.py": 100,
        "coop/commands/destroy.py": 69,
        "TOTAL": 90,
    },
    "run_mypy": False,
}


setup(
    name="coop",
    version="6.9",
    packages=["coop", "coop.commands"],
    entry_points={"console_scripts": ["coop = coop.coop:main"]},
    install_requires=["click", "tabulate"],
    test_suite=ptr_params["test_suite"],
)
"""

SAMPLE_PYPROJECT = """\
[tool.ptr]
disabled = true
entry_point_module = "ptr"
test_suite = "ptr_tests"
test_suite_timeout = 120
required_coverage = { 'ptr.py' = 84, TOTAL = 88 }
run_usort = true
run_black = true
run_mypy = true
run_flake8 = true
run_pylint = false
run_pyre = true
"""

# Disabled is set as we --run-disabled the run in CI
SAMPLE_SETUP_CFG = """\
[ptr]
disabled = true
entry_point_module = ptr
test_suite = ptr_tests
test_suite_timeout = 120
required_coverage_ptr.py = 84
required_coverage_TOTAL = 88
run_usort = true
run_black = true
run_mypy = true
run_flake8 = true
run_pylint = false
run_pyre = true
"""

SAMPLE_SETUP_PY_PTR = {
    "test_suite": "coop.tests.base",
    "test_suite_timeout": 60,
    "required_coverage": {
        "coop/coop.py": 99,
        "coop/commands/pwn.py": 100,
        "coop/commands/destroy.py": 69,
        "TOTAL": 90,
    },
    "run_mypy": False,
}
